swarm/storage/resource.go

```diff
@@ -417,44 +441,73 @@ func (self *ResourceHandler) loadResource(name string, refresh bool) (*resource,
 }
 
 // update mutable resource index map with specified content
-func (self *ResourceHandler) updateResourceIndex(rsrc *resource, chunk *Chunk, indexname *string) (*resource, error) {
-
-	// rsrc update data chunks are total hacks
-	// and have no size prefix :D
-	err := self.verifyContent(chunk.SData)
-	if err != nil {
-		return nil, err
+func (self *ResourceHandler) updateResourceIndex(rsrc *resource, chunk *Chunk) (*resource, error) {
+
+	// retrieve metadata from chunk data and check that it matches this mutable resource
+	signature, period, version, name, data, err := self.parseUpdate(chunk.SData)
+	if *rsrc.name != name {
+		return nil, fmt.Errorf("Update belongs to '%s', but have '%s'", name, *rsrc.name)
+	}
+	log.Trace("update", "name", *rsrc.name, "rootkey", rsrc.nameHash, "updatekey", chunk.Key, "period", period, "version", version)
+	// only check signature if validator is present
+	if self.validator != nil {
+		digest := self.keyDataHash(chunk.Key, data)
+		_, err = getAddressFromDataSig(digest, *signature)
+		if err != nil {
+			return nil, fmt.Errorf("Invalid signature: %v", err)
+		}
 	}
 
 	// update our rsrcs entry map
-	period, version, _, data, err := parseUpdate(chunk.SData[signatureLength:])
+	rsrc.lastKey = chunk.Key
 	rsrc.lastPeriod = period
 	rsrc.version = version
 	rsrc.updated = time.Now()
 	rsrc.data = make([]byte, len(data))
 	copy(rsrc.data, data)
-	log.Debug("Resource synced", "name", rsrc.name, "key", chunk.Key, "period", rsrc.lastPeriod, "version", rsrc.version)
-	self.setResource(*indexname, rsrc)
+	log.Debug("Resource synced", "name", *rsrc.name, "key", chunk.Key, "period", rsrc.lastPeriod, "version", rsrc.version)
+	self.setResource(*rsrc.name, rsrc)
 	return rsrc, nil
 }
 
-func parseUpdate(blob []byte) (period uint32, version uint32, ensname []byte, data []byte, err error) {
-	headerlength := binary.LittleEndian.Uint16(blob[:2])
-	if int(headerlength+2) > len(blob) {
-		return 0, 0, nil, nil, fmt.Errorf("Reported header length %d longer than actual data length %d", headerlength, len(blob))
+// retrieve update metadata from chunk data
+// mirrors newUpdateChunk()
+func (self *ResourceHandler) parseUpdate(chunkdata []byte) (*Signature, uint32, uint32, string, []byte, error) {
+	var err error
+	cursor := 0
+	headerlength := binary.LittleEndian.Uint16(chunkdata[cursor : cursor+2])
+	cursor += 2
+	datalength := binary.LittleEndian.Uint16(chunkdata[cursor : cursor+2])
```

Comment:
Rather than using a cursor, you could just decode the chunk directly into a struct using [`binary.Read`](https://golang.org/pkg/encoding/binary/#Read)

Reply:
I really have to get myself a hobby letting my write some C ;)

Reply:
Actually, the data is not fixed size, and the [https://golang.org/pkg/encoding/binary/#Read](docs) state:

```
func Read(r io.Reader, order ByteOrder, data interface{}) error
Read reads structured binary data from r into data. Data must be a
    pointer to a fixed-size value or a slice of fixed-size values.
```

So I guess I get to keep my nerdy cursor? :)

Nice to know about this, though!